feat(web): Media Library Feature

ollioddi opened the pull request:
# feat(web): Media Library Feature

## 1. Description

This PR introduces a complete **Media Library** feature for the Educado web application, providing a system for uploading, browsing, previewing, and managing media assets (images, videos, documents). The feature integrates seamlessly with Strapi's upload API and follows the established feature-based architecture.

### Highlights
- Full media management: upload, browse, preview, edit metadata, delete
- Drag and drop upload
- Rich metadata editing: filename, alt text, caption with validation
  - Allows editing metadata BEFORE **upload** or **selection**, so the asset is born with metadata. 
- Fullscreen image preview with zoom, rotate, and download capabilities
- Composable component architecture for flexible integration
- Form integration with React Hook Form
- Supports image (With full screen previews), videos (With inline player) and files (without preview)
  - Maybe at some point a content creator needs to share a pdf? 
- Uses responsive images in various sizes.
- Bulk delete option in overview page

> **Note**: This branch includes cherry-picked commits from `web/data-display-filters` for DataDisplay/table styling improvements that are dependencies for this feature.

![media](https://github.com/user-attachments/assets/157d5cf5-3ed1-4463-8f03-3fb6aaae0ce6)

---

## 2. What's Changed

### New: Media Feature (`web/src/features/media/`)

| Directory | Files | Purpose |
|-----------|-------|---------|
| `api/` | `media-mutations.tsx`, `media-queries.tsx` | TanStack Query mutations for upload/update/delete |
| `components/` | 12 components | Complete UI component suite (see Architecture) |
| `hooks/` | `use-file-upload.ts` | Custom hook for file upload with progress |
| `lib/` | `media-schemas.ts`, `media-utils.ts`, `media-columns.tsx` | Validation schemas, utilities, table columns |
| `pages/` | `media-overview-page.tsx` | Main media library page |

### Modified: Shared Components

| File | Change |
|------|--------|
| `form-file-upload.tsx` | Refactored to use new `MediaPickerTrigger` |
| `overlay-status-wrapper.tsx` | Added `minDuration` and `onSuccessComplete` props |
| `shadcn/accordion.tsx` | **New** - Added shadcn/ui accordion component |
| `shadcn/button.tsx`, `input.tsx`, etc. | Minor styling tweaks |

### Modified: DataDisplay (cherry-picked from `web/data-display-filters`)

- Enhanced `usePaginatedData` to support flat array responses
- Improved table styling to match Figma prototype
- Added filter dropdown and quick filters components

### Removed

| File | Reason |
|------|--------|
| `shared/components/file-upload.tsx` | Replaced by media feature components |
| `shared/hooks/use-file-upload.ts` | Moved to media feature |
| `shared/Table/*` | Unused legacy table components |

---

## 3. Architecture

```
media/
├── api/
│   ├── media-mutations.tsx    # Upload, update, delete mutations
│
├── components/
│   │
│   │ # PRIMITIVES (Building Blocks)
│   ├── MediaUploadCard        # Dashed dropzone card UI
│   ├── MediaAssetPreview      # Renders image/video/file preview
│   ├── MediaMetadataForm      # Filename, alt, caption fields
│   │
│   │ # COMPOSITES (Complete Features)  
│   ├── MediaUploadZone        # Full upload flow with metadata editing
│   ├── MediaPickerTrigger     # Click to open picker, shows selection
│   ├── MediaPickerModal       # Modal: Upload tab + Browse tab
│   │
│   │ # DISPLAY
│   ├── MediaCard              # Grid card with edit/delete actions
│   ├── MediaDisplayEditor     # DataDisplay wrapper for media library
│   └── ImageFullscreenPreview # Fullscreen viewer with controls
│
├── hooks/
│   └── use-file-upload.ts     # Upload hook with FormData handling
│
├── lib/
│   ├── media-schemas.ts       # Zod schemas & validation
│   ├── media-utils.ts         # Utilities (URL, bytes, mime types)
│   └── media-columns.tsx      # Table column definitions
│
└── pages/
    └── media-overview-page.tsx
```

### Component Hierarchy

```
MediaPickerTrigger (form integration)
└── MediaPickerModal
    ├── Tab: Upload
    │   └── MediaUploadZone
    │       ├── MediaUploadCard (dropzone)
    │       └── MediaFilePreviewItem[] (per file)
    │           ├── MediaAssetPreview
    │           └── MediaMetadataForm
    │
    └── Tab: Browse
        └── MediaDisplayEditor
            └── DataDisplay (grid of MediaCards)
                └── MediaCard
                    ├── MediaAssetPreview
                    └── ImageFullscreenPreview (on click)
```

### Validation

All media metadata is validated using Zod schemas:

```typescript
MEDIA_METADATA_LIMITS = {
  filename: 50,  // Cross-platform safe
  alt: 125,      // SEO best practice
  caption: 200,
}

// Filename validation includes:
// - Invalid characters: \ / : * ? " < > |
// - Windows reserved names: CON, PRN, AUX, NUL, COM1-9, LPT1-9
// - No leading/trailing spaces or periods
```

### Responsive images
Strapi can automatically genereate smaller versions of the same asset. This PR implements this as well.

The `media-asset-preview.tsx` defines an interface of the currently supported sizes.

```ts
interface StrapiImageFormats {
  large?: StrapiImageFormat;
  medium?: StrapiImageFormat;
  small?: StrapiImageFormat;
  thumbnail?: StrapiImageFormat;
}
```

Which is accepted as a param:

```
interface MediaAssetPreviewProps {
  asset: UploadFile;
  className?: string;
  imageClassName?: string;
  /** Whether to show controls for video */
  showVideoControls?: boolean;
  /** Whether to enable full screen preview for images */
  enableFullScreen?: boolean;
  /**
   * Size of the image to display.
   * Defaults to "medium" to optimize performance.
   */
  size?: MediaPreviewSize;
}
```

In this implementation:

- Asset grid is a "Small" size
- Table row is a "Thumbnail" size
- Selection preview is "Medium" size
- Asset fullscreen preview is "original" size

<img width="1461" height="462" alt="image" src="https://github.com/user-attachments/assets/dc3949d8-3d67-47bf-b469-c4553549daba" />


---

## 4. Usage

### Simple Upload Zone (Media Library Page)

```tsx
<MediaUploadZone
  fileTypes={["image", "video"]}
  maxFiles={10}
  onUploadComplete={(files) => {
    console.log('Uploaded:', files);
  }}
/>
```

### File Picker for Forms

```tsx
const [selectedFile, setSelectedFile] = useState<UploadFile | null>(null);

<MediaPickerTrigger
  value={selectedFile}
  onChange={setSelectedFile}
  fileTypes="image"
  maxFiles={1}
/>
```

### React Hook Form Integration

```tsx
import { FormFileUpload } from "@/shared/components/form/form-file-upload";

<FormFileUpload
  name="coverImage"
  control={form.control}
  label="Cover Image"
  fileTypes="image"
/>
```

### Fullscreen Preview (Standalone)

```tsx
<ImageFullscreenPreview
  isOpen={isPreviewOpen}
  onClose={() => setIsPreviewOpen(false)}
  imageUrl={asset.url}
  altText={asset.alternativeText}
  onDownload={() => downloadFile(asset)}
/>
```

---

## Screenshots
WIP SCREENSHOTS!!!

<img width="1844" height="1170" alt="image" src="https://github.com/user-attachments/assets/ff32ee13-d8f4-4db9-a6f8-808166276c0e" />

<img width="1829" height="1143" alt="image" src="https://github.com/user-attachments/assets/8ec39ffa-ca2d-4c39-803a-a160b3f3f4e4" />

<img width="2548" height="1193" alt="image" src="https://github.com/user-attachments/assets/0d7701b1-19fa-47aa-ab47-0bd8e8506bb9" />

### Video support
<img width="1833" height="749" alt="image" src="https://github.com/user-attachments/assets/6c08a3d9-2b3e-4b6b-87a5-d5adf35fba16" />

### Modal picker
Used for integration in forms. Allows the user to upload OR select an existing asset.
<img width="2505" height="1107" alt="image" src="https://github.com/user-attachments/assets/79b0e5f9-54a5-4d10-9ecc-9846d63dcf3f" />

<img width="2538" height="1207" alt="image" src="https://github.com/user-attachments/assets/ddab1f4a-b9a2-46b4-aac1-520ed7d01642" />

<img width="2510" height="1105" alt="image" src="https://github.com/user-attachments/assets/f79ba3c2-2b86-4886-b407-b16ddac9cdf1" />

### Bulk delete menu
The bulk removal UI replaces the preview of a single asset. This allows easy cleanup.

<img width="1131" height="814" alt="image" src="https://github.com/user-attachments/assets/1bb90aba-4d40-461e-9417-95da2818e3ab" />


---

## Checklist

- [x] Tested upload flow (single and multiple files)
- [x] Tested metadata editing (inline and modal)
- [x] Tested fullscreen preview controls
- [x] Tested form integration with React Hook Form
- [x] Tested file type filtering
- [x] Verified responsive design
- [x] No console errors/warnings

---

## Related

- Cherry-picks from: `web/data-display-filters` (5 commits for DataDisplay improvements)
- Depends on: Strapi Upload API (`/api/upload`)

Commit:
Border on select and multinput

Commit:
Small design adjustments

Commit:
Changed grid/list button to look like figma

MikkelBLarsen committed:
Added translation to rest of the english words

Alb9ert committed:
feat: (web) file upload to course (#29)

* start

* save

* fixed

* fix

ollioddi committed:
Test

ollioddi committed:
Test

Commit:
DataToolBar fix

+ more

ollioddi committed:
Feat(web): Multi-select add item callback & ref. Course Editor add new category (#31)

* multiselect new entries

can now add new entries in multiselects with a popup modal and error messages

* translation, button fix, course info multiselect

* Fix categories load, category creation. Clean error & loading states

* Translation fix

* feat: (web) file upload to course (#29)

* start

* save

* fixed

* fix

---------

Co-authored-by: Alfred <47451379+Alfenstein8@users.noreply.github.com>
Co-authored-by: Smidtty <148750415+Smidtty@users.noreply.github.com>
Co-authored-by: Albert <148211219+Alb9ert@users.noreply.github.com>

Commit:
Update course-card.tsx

Commit:
Update course-card.tsx

ollioddi committed:
Feat(Web): Auth Context (#30)

* Add useAuth hook, sync with language and preferences for usePaginatedData.

* Update web/src/shared/hooks/use-local-storage.ts

Co-authored-by: Albert <148211219+Alb9ert@users.noreply.github.com>

* Update web/src/shared/hooks/use-isomorphic-layout-effect.tsx

Co-authored-by: Albert <148211219+Alb9ert@users.noreply.github.com>

* Update web/src/shared/hooks/use-event-callback.tsx

Co-authored-by: Albert <148211219+Alb9ert@users.noreply.github.com>

---------

Co-authored-by: Alfred <47451379+Alfenstein8@users.noreply.github.com>
Co-authored-by: Albert <148211219+Alb9ert@users.noreply.github.com>

ollioddi committed:
Course editor menu

ollioddi committed:
Unchecked color

Alb9ert committed:
removed workflow for now

Alb9ert committed:
removed workflow for now

ollioddi committed:
Package lock

Alb9ert committed:
scripts

Alb9ert committed:
moved btn

Alb9ert committed:
ui fix

Alb9ert committed:
ready for backend

ollioddi committed:
Feat: 🚀 Add Turborepo, CI/CD Pipelines & Comprehensive Documentation (#59)

* Turborepo, npm scripts & docker

* Docs

* Update contributions docs

* Forgot to add Web to compose. Fix build errors too.

* Add and update docker docs.

* Check Workflow fix

* Experiment: Summary

* Test summary as comment

* Error handling

* Add permissions to workflow, to post message

* Update to reflect PR comment

ollioddi committed:
Fix npm not installing (#73)

* Fix npm not installing

ollioddi committed:
Update lockfile to test npm ci

Alb9ert committed:
delete lock file

ollioddi committed:
Fix: Restore OpenAPI types for client generation via Hey API + deprecated Strapi spec (temporary) (#68)

* Swap generator, update api-config for .env injection.

* Delete old client

* Generate new Hey API client

* Throw errors instead of returning them in the response.

* Fix ts error

* new functions for muutations (heyapi)

* Fix generation and implementations

* Fixes

---------

Co-authored-by: Smidtty <148750415+Smidtty@users.noreply.github.com>

ollioddi committed:
hm

ollioddi committed:
Foo

Alb9ert committed:
fix

ollioddi committed:
Test

ollioddi committed:
Full CI test with Strapi changes

ollioddi committed:
Test build

ollioddi committed:
CI: Test build

ollioddi committed:
WIP filters & merge

ollioddi committed:
Tweaks

Alb9ert committed:
fix

ollioddi committed:
(chore/web) cleanup dependencies (#70)

* Dependency cleanup

* Remove Yup

ollioddi committed:
data-table selection V1 (#78)

- Table has selection
- Grid has selection
- Callback for getting selected values
- Interface to limit selection

ollioddi committed:
0.2.0

ollioddi committed:
Seperate concerns

CelestrialDragon committed:
[US-0004] - Content Creator Login #53

* new route

Added a new rotue with is for login verification for content-creators

* login system

Login system no cryptation

* tab

* email verfication resposes

changed some errors

* fixed a bug using test

Created a http request to test the system
found a mistake and fixed it

* Backend works as it should

TO run the http Test you need to create a user with the email and password depictaed in the http

* Revert "Merge branch 'dev' into feature-ContentCreatorLogin"

This reverts commit 823cfd45de2ca5eba320e976983f490d38bf20dd, reversing
changes made to 003d93225fb3869410396ebe80822ea443202e7d.

* causel

causel fixe ui

* fixed lendning

* more eslint

* welcome

* Texst placed the right place

* documents

* update complete

* it works?

* begining to do stuff

* feedback from luise

* error handling

* Spelling mistake fix

* Tailwind login update

Added tailwind colors and fonts to login but some of the colors are wrong  so that has to be fixed.

* npm run

* new route

Added a new rotue with is for login verification for content-creators

* login system

Login system no cryptation

* tab

* email verfication resposes

changed some errors

* fixed a bug using test

Created a http request to test the system
found a mistake and fixed it

* Backend works as it should

TO run the http Test you need to create a user with the email and password depictaed in the http

* Revert "Merge branch 'dev' into feature-ContentCreatorLogin"

This reverts commit 823cfd45de2ca5eba320e976983f490d38bf20dd, reversing
changes made to 003d93225fb3869410396ebe80822ea443202e7d.

* causel

causel fixe ui

* fixed lendning

* more eslint

* welcome

* Texst placed the right place

* documents

* update complete

* it works?

* begining to do stuff

* feedback from luise

* error handling

* Spelling mistake fix

* Tailwind login update

Added tailwind colors and fonts to login but some of the colors are wrong  so that has to be fixed.

* npm run

* Restore missing files from origin/dev

* Foo

* content creator stuff

* no white screen

* it works

* last bit changes

* eslist

* figma things that we forgot

* test and found an error

* some add notes

* add co uthor

Co-Authored-By: Sozh03 <152170949+Sozh03@users.noreply.github.com>

* Delete strapi/src/api/content-creator/content-creator.http

* errorCodes.ts

in responds to comment by Cathredral101

Co-Authored-By: Sozh03 <152170949+Sozh03@users.noreply.github.com>

* jwt key

update based on comment

Co-Authored-By: Sozh03 <152170949+Sozh03@users.noreply.github.com>

* Update content-creator.ts

Co-Authored-By: Sozh03 <152170949+Sozh03@users.noreply.github.com>

* thisngs that some how got deleted

Co-Authored-By: Sozh03 <152170949+Sozh03@users.noreply.github.com>

* shouldn't have been changed

Co-Authored-By: Sozh03 <152170949+Sozh03@users.noreply.github.com>

* deleted per request

* type check content creator

---------

Co-authored-by: Sooonnnaaa <Bubulspam@gmail.com>
Co-authored-by: Oliver Roed Schøler <oliver@roedscholer.dk>
Co-authored-by: Sozh03 <152170949+Sozh03@users.noreply.github.com>
Co-authored-by: Johny1511 <83210690+Johny1511@users.noreply.github.com>

EmilLykke committed:
User relation

EmilLykke committed:
Updated

ollioddi committed:
fix: Update Swagger to use proper merged spec (#90)

* feat: enhance API documentation and update security policies

- Added OpenAPI documentation HTML file and .gitkeep for the public/openapi directory.
- Updated middlewares to include a detailed content security policy.
- Modified package scripts to ensure the OpenAPI spec is copied during development and build processes.

* Update scripts to new workflow.

* Fix dependency

EmilLykke committed:
remove relation

EmilLykke committed:
updated return type and added policies to routes

ollioddi committed:
Bunch of cleanup.

EmilLykke committed:
updated some endpoints to not require auth

EmilLykke committed:
updated strapi spech with new definitions for endpoints

EmilLykke committed:
small change to registartion verifiedAt

EmilLykke committed:
updated content creator login function name change

EmilLykke committed:
real generated

EmilLykke committed:
Login.tsx working

EmilLykke committed:
updated strapi

EmilLykke committed:
Strapi login for content creators

EmilLykke committed:
Small update

EmilLykke committed:
Overwrite upload route

EmilLykke committed:
small changes

EmilLykke committed:
PR comments

edmxa committed:
Add required durationHours field to course schema

Introduces a required decimal field 'durationHours' to the course content type with a minimum value of 0.1. Updates OpenAPI and documentation specs to reflect the new field, including its presence in required properties and type definitions. Also adds JWTContentCreatorResponse schema to OpenAPI spec.

ollioddi committed:
feat: Add minDuration to OverlayStatusWrapper and onSuccessComplete callback to properly animate. Updated consumers of it.

ollioddi committed:
Linting

ollioddi committed:
fix: usePaginatedData supports flat array responses from Strapi

ollioddi committed:
add shadcn accordion

CelestrialDragon committed:
Feature/tailwind for login (#113)

* testing

* updating hex codes to tailwind

ollioddi committed:
WIP

DandingB committed:
Fix/notifications (#103)

* Changed notification dropdown to use shadcn popover

* Changed notification toast to use the shadcn toast

* Update course-editor-review.tsx

* Removed toasts from notifications

* Update course-editor-review.tsx

* Added translations

* Fixed imports

---------

Co-authored-by: DandingB <daniel@Daniels-MBP-2.home>
Co-authored-by: DandingB <daniel@Daniels-MacBook-Pro-2.local>

edmxa committed:
Make lastName required for Content Creator (#122)

Updated the Content Creator schema and related OpenAPI and TypeScript definitions to require the lastName field. This ensures lastName is mandatory for content creator entities and improves validation consistency across the API and documentation.

ollioddi committed:
Cleanup and fixes.

Smidtty committed:
Migration of all icons to a singular library (mdi) (#100)

* removed hero-icons and react-icons

* exchanged lucide icons with mdi

EmilLykke commented on web/src/features/media/api/media-queries.tsx:
This would be fixed/deleted before merge?

EmilLykke reviewed (commented):
I haven't looked through all but here there is some comments

Fahnoee committed:
updated package lock (#126)

PeterDunlop committed:
Added a (pre)vious slide function that was missing for some reason (#127)

RasmusJensen1 committed:
51us0045  - update profile creator (name and bio) (#112)

* Update profile handling and integrate content creator API

* fix: Update profile to make password optional and handle empty values - due to strapi hasing password and then overwriting the the exiting password with the hashed one - very sad

* PR comments fixed

fix: Update profile to make password optional and improve form handling

* PR comment fix:

fix: update import path for staticForm component and add staticForm functionality

Johny1511 committed:
[US-0091] - Feature dashboard atividades functionality (#115)

* dashboard activities with custom get route

* Front-end display of dashboard activities

* Update custom-dashboard-activity.ts

Sorting the activities on creation date now.

* Update course-overview-sidebar.tsx

unused dependency

* dashboard activities with custom get route

* Front-end display of dashboard activities

* Update custom-dashboard-activity.ts

Using map function for cleaner code (pr review request)

* Update custom-dashboard-activity.ts

User extraction is now a helper function, request from PR

ollioddi committed:
feat: data-display filtering (#99)

* WIP filters & merge

* Tweaks

* Removed unused table

* Update table styling to match figma prototype

* Seperate concerns

* Small fixes

* Bunch of cleanup.

* Cleanup

* Remove selection

* Reuse function

* Fix assertion

* Remove null check

* Disable logging by default

* Revert bool

* Fix linting errors.

ollioddi committed:
Deployment fixes (#133)

* Deployment fixes

* Fix broken test-page

* Updated strapi & fix dockerfiles

* Update package-lock

---------

Co-authored-by: mewkel <mikkelbrunl@gmail.com>

ollioddi committed:
Fix and cleanup translations, other tweaks.

ollioddi committed:
Remove uses of eslint-disable

ollioddi committed:
ESLint fixes

ollioddi committed:
Consistent card height and width.

ollioddi committed:
Install deps

ollioddi committed:
Test

ollioddi committed:
Update to mdi?

ollioddi committed:
Add bulk delete option

PeterDunlop commented on web/src/features/media/api/media-mutations.tsx:
```diff
@@ -0,0 +1,175 @@
+        queryKey: [["media"]],
+        exact: false,
+        type: "active",
+      });
```
```suggestion
      refetchMedia(queryClient);
```

PeterDunlop commented on web/src/features/media/api/media-mutations.tsx:
```diff
@@ -0,0 +1,175 @@
+        queryKey: [["media"]],
+        exact: false,
+        type: "active",
+      });
```
```suggestion
       refetchMedia(queryClient);
```

ollioddi committed:
Reject unsupported filetype from drag-n-drop

ollioddi committed:
Allow file uploads in overview page.

ollioddi committed:
Remove react-scan

ollioddi commented:
> Currently the program allows you to drag and drop a different file type (Such as a txt file) which causes the program to crash. Please look into this.

This has been fixed. Also fixed an issue where selecting via the file browser limits to the accepted type of files, but all validation where stripped from dragging a file.

Dragging an unsupported file now doesnt upload it, and shows a notification that the upload was cancelled.

ollioddi committed:
Add tooltips

ollioddi committed:
Create reusable refetchMedia

ollioddi committed:
Null checks

ollioddi commented on web/src/features/media/components/media-upload-card.tsx:
```diff
@@ -0,0 +1,139 @@
+      // Reset input so same file can be selected again
+      e.target.value = "";
+    }
+  };
```
It is slightly enforced by the input:

`multiple={allowMultiple}`, but ill add a check for the actual limit.

ollioddi committed:
Add error handling & file limit validation

ollioddi committed:
Revert uninteded rebase changes.

PeterDunlop commented:
Sonarqube has passed all tests on this, yet there are still changes that can be made in the code based on Sonarqube suggestions? Is this some kind of bug, or are the remaining Sonarqube changes unnecessary?

HasnainAliprogramming reviewed (commented):
Godt arbejde med denne feature. Jeg har været hele PR’en igennem, og der er meget, der fungerer godt. Strukturen i media-modulet er fornuftig, og flowet mellem upload, gallery og preview hænger sammen. Der er dog nogle ting, der skal rettes, før vi kan merge.

Upload-valideringen skal være mere robust. Lige nu bliver almindelige fejl håndteret, men vi mangler stadig at dække spoofede filendelser, mismatch mellem MIME-type og det faktiske indhold og en mere tydelig håndtering af backend-fejl, når en fil bliver afvist. Frontend og backend skal reagere ens, ellers får vi tekniske edge cases senere. Det er vigtigt, at det spiller helt stabilt, da det er en central del af modulet.

Import-fejlen omkring media-queries.tsx skal løses. Enten mangler filen, eller også er importstien forkert. Uanset hvad må builden ikke fejle, så den skal rettes, før vi kan komme videre.

Jeg kan også se, at thumbnails nogle steder loader større billedefiler end nødvendigt. Det skal optimeres, så vi ikke trækker tunge assets ind i gridet. Preview-komponenten er desuden ret tung og bør lazy-loades, så den ikke påvirker initial load, når brugeren ikke er i nærheden af previewet.

Metadata-flowet virker fint, men der bør tilføjes en simpel validering, så tom alt-text eller andre tomme felter ikke bliver gemt ved en fejl. Det er en lille ændring, men det forbedrer datakvaliteten betydeligt.

Der mangler desuden en kort README i features/media, som forklarer brugen af MediaPicker, de vigtigste props og de hooks, der skal bruges. Det gør modulet nemmere at arbejde med for andre i teamet, og det sparer tid på sigt.

Til sidst vil jeg anbefale, at vi får lavet tests . Upload-flow, fejlscenarier, preview og metadata bør dækkes, da det her modul får en stor rolle i brugeroplevelsen, og vi skal undgå regressioner fremover.

Når upload-valideringen er justeret, import-fejlen er rettet, thumbnail-loading er optimeret, preview lazy-load er på plads, metadata-valideringen er tilføjet, og CI kører grønt, så er vi klar til at merge. det var lige det jeg kunne komme med, Rigtig godt arbejde med det hele.

ollioddi committed:
Update web/src/features/media/lib/media-utils.ts

Co-authored-by: Peter Simon Wallace Dunlop <90600834+PeterDunlop@users.noreply.github.com>

ollioddi commented:
> Godt arbejde med denne feature. Jeg har været hele PR’en igennem, og der er meget, der fungerer godt. Strukturen i media-modulet er fornuftig, og flowet mellem upload, gallery og preview hænger sammen. Der er dog nogle ting, der skal rettes, før vi kan merge.
> 
> Upload-valideringen skal være mere robust. Lige nu bliver almindelige fejl håndteret, men vi mangler stadig at dække spoofede filendelser, mismatch mellem MIME-type og det faktiske indhold og en mere tydelig håndtering af backend-fejl, når en fil bliver afvist. Frontend og backend skal reagere ens, ellers får vi tekniske edge cases senere. Det er vigtigt, at det spiller helt stabilt, da det er en central del af modulet.
> 
> Import-fejlen omkring media-queries.tsx skal løses. Enten mangler filen, eller også er importstien forkert. Uanset hvad må builden ikke fejle, så den skal rettes, før vi kan komme videre.
> 
> Jeg kan også se, at thumbnails nogle steder loader større billedefiler end nødvendigt. Det skal optimeres, så vi ikke trækker tunge assets ind i gridet. Preview-komponenten er desuden ret tung og bør lazy-loades, så den ikke påvirker initial load, når brugeren ikke er i nærheden af previewet.
> 
> Metadata-flowet virker fint, men der bør tilføjes en simpel validering, så tom alt-text eller andre tomme felter ikke bliver gemt ved en fejl. Det er en lille ændring, men det forbedrer datakvaliteten betydeligt.
> 
> Der mangler desuden en kort README i features/media, som forklarer brugen af MediaPicker, de vigtigste props og de hooks, der skal bruges. Det gør modulet nemmere at arbejde med for andre i teamet, og det sparer tid på sigt.
> 
> Til sidst vil jeg anbefale, at vi får lavet tests . Upload-flow, fejlscenarier, preview og metadata bør dækkes, da det her modul får en stor rolle i brugeroplevelsen, og vi skal undgå regressioner fremover.
> 
> Når upload-valideringen er justeret, import-fejlen er rettet, thumbnail-loading er optimeret, preview lazy-load er på plads, metadata-valideringen er tilføjet, og CI kører grønt, så er vi klar til at merge. det var lige det jeg kunne komme med, Rigtig godt arbejde med det hele.

Det havde været nemmere at besvare med et traditionelt review, men her kommer den:

**1. "Upload-valideringen skal være mere robust. Lige nu bliver almindelige fejl håndteret, men vi mangler stadig at dække spoofede filendelser, mismatch mellem MIME-type og det faktiske indhold og en mere tydelig håndtering af backend-fejl, når en fil bliver afvist. Frontend og backend skal reagere ens, ellers får vi tekniske edge cases senere."**

Forstår ikke helt hvordan den skal være mere robust? Filtype tjek laves af input elementerne, selv på drag-n-drop. Spoofede filendelser er ærligt talt ude af scope i denne PR, og nok ikke reelt et problem da det kræver login at kunne uploade.

Derudover er det jo et almindeligt API endpoint, så jeg antager at Strapi også laver deres på alle filer. 

Hvad betyder en mere tydelig håndtering af backend fejl?

Hvis en fejl opstår i mutationen vil den blive vist i UI.

**2. "Import-fejlen omkring media-queries.tsx skal løses"**

Jeg har ingen referencer til media-queries. Filen er slettet? Og buildet fejler ikke, mit seneste push har passet alt CI?

**3. "Jeg kan også se, at thumbnails nogle steder loader større billedefiler end nødvendigt."**

Det er by design? Jeg har skrevet i PR hvornår hvilke størrelser anvendes, og det virkede passende ift. visuel kvalitet på en større skærm. Altså hvis du har noget konkret, kan jeg med ét argument ændre hvilken en som vises, men er ikke helt sikker på hvor du mener?

**4. " Preview-komponenten er desuden ret tung".** 
Hvilken en af dem?

**5. "Metadata-flowet virker fint, men der bør tilføjes en simpel validering".** 

Den har validering?

Det skal også være muligt at slette og nulstille. Desuden tillader jeg ikke ændring af fil-extension. Det er med vilje lavet til ikke at kunne redigeres, se tilbage på **punkt 1**. også ift. nedenstående:

```
const mediaAssetFormSchema = z.object({
  filename: filenameSchema,
  alt: z
    .string()
    .max(
      MEDIA_METADATA_LIMITS.alt,
      `Alt text must be less than ${String(MEDIA_METADATA_LIMITS.alt)} characters`
    )
    .optional(),
  caption: z
    .string()
    .max(
      MEDIA_METADATA_LIMITS.caption,
      `Caption must be less than ${String(MEDIA_METADATA_LIMITS.caption)} characters`
    )
    .optional(),
});

/** Character limits for media metadata fields */
export const MEDIA_METADATA_LIMITS = {
    filename: 50,
    alt: 125,
    caption: 200,
} as const;

// ============================================================================
// Filename Validation
// ============================================================================

/**
 * Characters that are invalid in filenames across Windows, macOS, and Linux.
 * - Windows: \ / : * ? " < > |
 * - macOS/Linux: / and null character
 * We use the union of all to ensure cross-platform compatibility.
 */
const INVALID_FILENAME_CHARS = /[\\/:*?"<>|]/;

/**
 * Reserved filenames in Windows (case-insensitive).
 */
const WINDOWS_RESERVED_NAMES = /^(CON|PRN|AUX|NUL|COM[1-9]|LPT[1-9])$/i;

/**
 * Validates that a string is a valid filename (without extension).
 * 
 * Rules:
 * - Must not be empty or only whitespace
 * - Must not contain invalid characters: \ / : * ? " < > |
 * - Must not be a Windows reserved name (CON, PRN, AUX, NUL, COM1-9, LPT1-9)
 * - Must not start or end with a space or period
 * - Maximum length of 50 characters
 */
export const isValidFilename = (name: string): boolean => {
    if (name.length === 0 || name.trim().length === 0) return false;
    if (name.length > MEDIA_METADATA_LIMITS.filename) return false;
    if (INVALID_FILENAME_CHARS.test(name)) return false;
    if (WINDOWS_RESERVED_NAMES.test(name)) return false;
    if (name.startsWith(" ") || name.endsWith(" ")) return false;
    if (name.startsWith(".") || name.endsWith(".")) return false;
    return true;
};
```

**6. "Til sidst vil jeg anbefale, at vi får lavet tests"**

Jeg er helt enig, men det er der ærlig talt ikke tid til. Så hvis det er et krav, kommer jeg ikke til at merge. Vi har snart code-freeze også.

Og til sidst, så kører CI fint "og CI kører grønt".

Men tak for at kigge det igennem :D

ollioddi committed:
Refine error handling

ollioddi committed:
Reuse existing schemas and improve

ollioddi committed:
Fix

PeterDunlop reviewed (dismissed):
Everything looks good. Major Sonarqube issues have been solved such as cyclomatic complexity and suggestions have been accounted for and the code runs. Great job!!!

ollioddi committed:
Documentation & final code cleanup

Aw1221 committed:
fix #123. Fix admin panel to be aligned with figma

ollioddi committed:
Add media to navbar

ollioddi commented:
@HasnainAliprogramming @PeterDunlop I added docs in latest commit and cleaned up some small code. 

I also just added it to the new navbar which were recently merged.
<img width="1330" height="200" alt="image" src="https://github.com/user-attachments/assets/62891375-ce28-4f91-ac8a-b11df5f5603a" />

I wont push more changes now, so final approval?

[DOCS](https://github.com/ErasmusEgalitarian/educado-backend/blob/web/media/docs/web/features/media-library.md)

ollioddi committed:
Merge branch 'dev' into web/media

ollioddi committed:
Sonarqube fixes

ollioddi committed:
Fix

ollioddi committed:
Fix translations

PeterDunlop reviewed (approved):
Everything looks good now. The last issues with admin will be fixed in the PR that fixes admin panel.

JeebleJuice reviewed (approved):
Looks really nice, I have found no issues, when testing it
